Issue #2: Add New Endpoint for Historical Prices

asmi38 opened the pull request:
**Goal:**
Create a new endpoint to FUSE that will allow users to query for historical prices.

**Inputs:**
Symbols (required): the stock ticker symbols that historical price data will be queried for
Date (not required, default = today): the date at which to get the historical prices for
Ranges (not required, default = 1 month): the range from the date to get the historical prices for

Note: if date and range are both present, date will supersede the range

**Function:**
IexRestController will serve as the endpoint for users of FUSE.

The method will call iexService which will have a new method getHistoricalPrices, this method will utilize the iexClient to fetch the data from the api.

The IexService getHistoricalPrices method will contain logic to call the correct method from the IexCloudClient depending on the amount of arguments passed in.

The IexCloudClient will be a new interface which will connect and make calls  to the Iex Cloud api. It will contain multiple methods which will be called by IexService.

The return value will be a list of IexHistoricalPrice objects which will contain the specified data for each day.

**Tests:**
The unit tests that will be conducted will check to see if the historicalData lines up with known dates and values and will work for varying inputs:

- Symbol, no date, no range
- Symbol, date, no range
- Symbol, no date, range
- No symbol input

asmi38 committed:
Issue #2: Add new Endpoint for Historical Prices

Create a new method in IexRestController getHistoricalPrice.

Inputs:
Symbols (required): the stock ticker symbols that historical price data will be queried for
Date (not required, default = today): the date at which to get the historical prices for
Ranges (not required, default = 1 month): the range from the date to get the historical prices for

IexRestController will serve as the endpoint for users of FUSE.

The method will call iexService which will have a new method getHistoricalPrices, this method will utilise the iexClient to fetch the data from the api.

The IexService getHistoricalPrices method will contain logic to call the correct method from the IexCloudClient depending on the amount of arguments passed in.

The IexCloudClient will be a new interface which will connect and make calls  to the Iex Cloud api. It will contain multiple methods which will be called by IexService.

The return value will be a list of IexHistoricalPrice objects which will contain the specified data for each day.

grundygolem2 commented on src/main/java/org/galatea/starter/service/IexClient.java:
```diff
@@ -1,6 +1,7 @@
 package org.galatea.starter.service;
 
 import java.util.List;
+import org.galatea.starter.domain.IexHistoricalPrice;
```
Don't import classes or methods you don't use

grundygolem2 commented on src/test/java/org/galatea/starter/entrypoint/IexRestControllerTest.java:
```diff
@@ -81,4 +81,88 @@ public void testGetLastTradedPriceEmpty() throws Exception {
+  public void testGetHistoricalPriceSymbol() throws Exception {
+
+    MvcResult result = this.mvc.perform(
+        org.springframework.test.web.servlet.request.MockMvcRequestBuilders
```
Can you avoid having the full path here?

grundygolem2 commented on src/main/java/org/galatea/starter/service/IexService.java:
```diff
@@ -45,5 +47,27 @@
+      final String range,
+      final String date) {
+    if (symbol.isEmpty()) {
+      return Collections.emptyList();
```
Given that we expect a symbol in our endpoint, is emptyList the correct response here?
(not being snarky, genuine question)

asmi38 replied:
This was to catch the scenario where a request is made that has the symbol parameter, but the value is empty, for example, this testcase:
```
        org.springframework.test.web.servlet.request.MockMvcRequestBuilders
            .get("/iex/historicalPrice?symbol=") 
```
It would satisfy the required condition of the @RequestParam symbol parameter in the IexRestController, but would be unable to make a successful request to the IEX API as the symbol is a required parameter.

A few alternative solutions would be to make the if elses more explicit and to return an emptyList as a base return case. 

I could also make the symbol @RequestParam not required and test for null but I don't think that would be an effective work around.

Another option would be to do the check in the IexRestController but I think it would be best to keep the logic in the Service class.

With this in mind, do you think I should still change it? Or perhaps there's a better alternative I'm missing

grundygolem2 replied:
Gotcha, that makes sense. In that case, you could look into adding request validation, because symbol= isn't really a valid request from the client. There are annotation based methods you could use to verify that symbol is populated.

asmi38 committed:
Updated Historical Price endpoint based on feedback

Added annotation validation to IexrestController that will respond with a bad request if symbol is empty. Updated IexService logic to remove the handling for this condition as it will no longer reach it.

Additionally, changed testGetHistoricalPriceEmpty to reflect these changes.

Removed unused imports in IexClient and IexCloudClient.

Updated IexRestControllerTest to reduce the path for MockMvcRequestBuilders.

grundygolem2 reviewed (approved):
Looks Good

alex-power commented on src/test/java/org/galatea/starter/entrypoint/IexRestControllerTest.java:
```diff
@@ -74,11 +75,94 @@ public void testGetLastTradedPrice() throws Exception {
+            .accept(MediaType.APPLICATION_JSON_VALUE))
+        .andExpect(status().isBadRequest())
+        .andReturn();
+  }
```
Please add one more test to validate the case if the IEX server is down or fails.

alex-power reviewed (changes requested):
Nice work. Please add one more test case per my comment.

asmi38 committed:
Added test and exception handling for IEX server errors

Added exception handling in the RestExceptionHandler class for exceptions in the FeignClient.

Added a test and associated mapping for a response status of 500. Replicating what would happen if the IEX server was down.

alex-power commented on src/main/java/org/galatea/starter/entrypoint/RestExceptionHandler.java:
```diff
@@ -73,8 +74,18 @@
 
+  @ExceptionHandler(FeignException.class)
+  protected ResponseEntity<Object> handleFeignStatusException(final FeignException exception) {
+    log.debug("Error connecting to IEX", exception);
```
Errors are not debug statements, should be ```log.error```

asmi38 committed:
Updated log

Changed from log.debug to log.error for FeignExceptions

asmi38 committed:
Merge branch 'develop' into Issue-2-New-Endpoint-For-Historical-Prices
